### PR TITLE
Shrink and Perturb Weight Initialization

### PR DESCRIPTION
**Pull request by melo-gonzo**
# PhysicsNeMo Pull Request

## Adds `physicsnemo.nn.shrink_and_perturb_`

In-place shrink-and-perturb weight re-initialization, [reference paper](https://arxiv.org/pdf/1910.08475):

`θ ← shrink·θ + perturb·ε`

- `θ`: pretrained model weights
- `shrink`: retention factor in `[0, 1)`, how much of each pretrained weight to keep
- `perturb`: noise level, scaled per-layer by the pretrained weight's std (default) or by a fresh init's std
- `ε`: random noise (Gaussian by default)

**Why.** Warm-starting from pretrained weights can converge to a *worse* loss asymptote
than training from scratch (a plasticity loss associated with grown weight norms).
Shrinking each weight toward zero and adding noise is intended to restore the scale
statistics and plasticity of a fresh init while keeping the pretrained direction,
helping recover that warm-start handicap.

**Use.**
```python
from physicsnemo.nn import shrink_and_perturb_
model.load_state_dict(pretrained)
shrink_and_perturb_(model, shrink=0.5, perturb=0.1)
```
Model-agnostic (any torch.nn.Module, including physicsnemo.Module); in place; buffers
untouched; tied params handled.

Noise (noise=). "scaled_normal" (default): ε = std(θ)·N(0,1), scaled per-tensor
by the pretrained weight's own spread. "normal": unit Gaussian. Or a callable handed
fresh scratch (shape/dtype/device), so torch.randn_like, torch.nn.init.*_, or a
custom sampler all work, e.g. scaling noise by a fresh init's std instead of the
pretrained one.

Applications. Transfer/fine-tuning from a pretrained checkpoint, continual and
active learning (to help re-inject plasticity between rounds), and other warm-start
settings where a fresh baseline might otherwise win.

Limitations. Intended as a repair for a bad warm start rather than a way to add
signal: it tends to interpolate between fine-tune and scratch. scaled_normal adds no noise
to scalar or constant tensors (std=0), and shrink/perturb generally need tuning.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/nn/module/utils/weight_init.py [hunk 1]
```diff
@@ -39,6 +41,237 @@ def trunc_normal_(*args, **kwargs):
     return _torch_trunc_normal_(*args, **kwargs)
 
 
+_NOISE_PRESETS = ("scaled_normal", "normal")
+
+
+def _resolve_noise(
+    noise: Union[
```
Comment (laserkelvin) [inline physicsnemo/nn/module/utils/weight_init.py:49]: This formatting looks kind of weird to me - is this actually ruff'd?

File: physicsnemo/nn/module/utils/weight_init.py [hunk 2]
```diff
@@ -39,6 +41,237 @@ def trunc_normal_(*args, **kwargs):
...
+        def make_eps(p: torch.Tensor) -> torch.Tensor:
+            return noise(torch.empty_like(p))
+
+    elif noise == "scaled_normal":
+
+        def make_eps(p: torch.Tensor) -> torch.Tensor:
+            if p.numel() < 2:
+                # std is undefined for a single element; shrink only, no noise.
+                return torch.zeros_like(p)
+            z = torch.empty_like(p).normal_(generator=generator)
```
Comment (laserkelvin) [inline physicsnemo/nn/module/utils/weight_init.py]: Could you not just do `randn_like`?

File: physicsnemo/nn/module/utils/weight_init.py [hunk 3]
```diff
@@ -39,6 +41,237 @@ def trunc_normal_(*args, **kwargs):
...
+def shrink_and_perturb_(
+    module: torch.nn.Module,
+    shrink: float = 0.5,
+    perturb: float = 0.1,
+    *,
+    noise: Union[
+        Literal["scaled_normal", "normal"], Callable[[torch.Tensor], torch.Tensor]
+    ] = "scaled_normal",
+    include: Optional[Callable[[str, torch.Tensor], bool]] = None,
+    generator: Optional[torch.Generator] = None,
```
Comment (laserkelvin) [inline physicsnemo/nn/module/utils/weight_init.py:122]: Do you know how `torch.manual_seed` propagates to this?

File: physicsnemo/nn/module/utils/weight_init.py [hunk 4]
```diff
@@ -39,6 +41,237 @@ def trunc_normal_(*args, **kwargs):
...
+        Predicate ``(name, param) -> bool`` selecting which parameters to
+        perturb. Parameters for which it returns ``False`` are left entirely
+        unchanged (not even shrunk). Default: all parameters. For warm-starting,
+        pass e.g. ``include=lambda n, p: n in transferred`` to perturb only the
+        transferred backbone. Tied parameters are visited under every alias so
+        the predicate can match any checkpoint key, but each underlying tensor
+        is updated at most once. A warning is issued if an explicit ``include``
+        matches no parameters.
+    generator : torch.Generator, optional
+        Generator for the built-in Gaussian noise, for reproducibility. Must be
```
Comment (laserkelvin) [inline physicsnemo/nn/module/utils/weight_init.py:191]: We don't validate the `generator` being on the same device as `module`, if it is passed

File: physicsnemo/nn/module/utils/weight_init.py [hunk 5]
```diff
@@ -39,6 +41,237 @@ def trunc_normal_(*args, **kwargs):
...
+    Intended warm-start workflow: load the pretrained weights, apply this to the
+    transferred parameters, then (re)create the optimizer, and only afterwards
+    ``torch.compile`` / wrap in DDP or FSDP. Run it *before* distributed
+    wrapping: under FSDP2 the per-tensor ``std`` of ``"scaled_normal"`` reduces
+    over a sharded parameter and triggers an all-gather.
+
+    ``"scaled_normal"`` is a deliberately model-agnostic variant: it scales the
+    noise by the *pretrained tensor's* own standard deviation rather than by a
+    freshly re-initialized network's per-layer initializer variance (the form
+    in Ash & Adams, §4). It needs no architecture knowledge and matches the
```
Comment (laserkelvin) [inline physicsnemo/nn/module/utils/weight_init.py]: Complete reference somewhere?

**Review by laserkelvin — CHANGES_REQUESTED** [review]
Few minor things to address

File: physicsnemo/nn/module/utils/weight_init.py (hunk 5 above)
Comment (melo-gonzo) [reply physicsnemo/nn/module/utils/weight_init.py]: added an arxiv link to the main reference area [a8e28d6](https://github.com/NVIDIA/physicsnemo/pull/1839/commits/a8e28d63832b190795fe2d96c497e0dc273bdc65)

File: physicsnemo/nn/module/utils/weight_init.py (hunk 2 above)
Comment (melo-gonzo) [reply physicsnemo/nn/module/utils/weight_init.py]: yup, would be more readable with `randn_like` as well. [7ea0018](https://github.com/NVIDIA/physicsnemo/pull/1839/commits/7ea00186d48ab5e8d989970433d408a482d161a2)

File: physicsnemo/nn/module/utils/weight_init.py (hunk 4 above)
Comment (melo-gonzo) [reply physicsnemo/nn/module/utils/weight_init.py:191]: explicit checking of device for generator and module params [7bd8b09](https://github.com/NVIDIA/physicsnemo/pull/1839/commits/7bd8b0907f1c63d03c17684ef21837165708707a)

File: physicsnemo/nn/module/utils/weight_init.py (hunk 1 above)
Comment (melo-gonzo) [reply physicsnemo/nn/module/utils/weight_init.py:49]: it is in fact ruff'd

File: physicsnemo/nn/module/utils/weight_init.py (hunk 3 above)
Comment (melo-gonzo) [reply physicsnemo/nn/module/utils/weight_init.py:122]: `torch.manual_seed` and `generator=none` will use a global generator state everywhere. if you use a specific `generator` in shrink and perturb, that specific generator will be different from what is used globally.

File: physicsnemo/nn/module/utils/weight_init.py [hunk 6]
```diff
@@ -237,6 +239,18 @@ def shrink_and_perturb_(
             f"shrink={shrink}, perturb={perturb}"
         )
 
+    # The preset samplers draw on `generator`, which must sit on the device of
+    # every parameter it seeds (a callable ignores it, and perturb == 0 draws
+    # nothing, so skip the check in those cases).
+    if generator is not None and not callable(noise) and perturb != 0.0:
```
Comment (laserkelvin) [inline physicsnemo/nn/module/utils/weight_init.py:258]: If the model is fully sharded, then, what's the expected behavior here then?

If it's out of scope, we just need to raise an exception and document it, but just wanted to make sure that we cover this instance

**Review by laserkelvin — APPROVED** [review]
Approved for the single worker case, but see my comment

File: physicsnemo/nn/module/utils/weight_init.py (hunk 6 above)
Comment (melo-gonzo) [reply physicsnemo/nn/module/utils/weight_init.py:258]: This would fail with a shareded model anyways. Adding a check and informative message. [0611a41](https://github.com/NVIDIA/physicsnemo/pull/1839/commits/0611a41752cf5b451f542332274720f441e18a46)

**Review by loliverhennigh — APPROVED** [review]
LGTM
